src/dom.js

```diff
@@ -214,10 +214,31 @@ export function rootNodeFor(node) {
     return node.getRootNode() || node;
   }
   let n;
-  for (n = node; !!n.parentNode; n = n.parentNode) {}
+  for (n = node; !!n.parentNode && !isShadowRoot(n); n = n.parentNode) {}
   return n;
 }
 
+/**
+ * Determines if value is actually a `ShadowRoot` node.
+ * @param {*} value
+ * @return {boolean}
+ */
+export function isShadowRoot(value) {
+  // TODO(#22733): remove in preference to dom's `rootNodeFor`.
+  if (!value) {
+    return false;
+  }
+  // Node.nodeType == DOCUMENT_FRAGMENT to speed up the tests. Unfortunately,
+  // nodeType of DOCUMENT_FRAGMENT is used currently for ShadowRoot nodes.
+  if (value.tagName == 'I-AMPHTML-SHADOW-ROOT') {
+    return true;
+  }
+  return (
+    value.nodeType == /* DOCUMENT_FRAGMENT */ 11 &&
+    Object.prototype.toString.call(value) === '[object ShadowRoot]'
+  );
```

Comment:
Do we need this code path? I think we can just check `tagName` inline and call it a day.

Reply:
I think so. It's too difficult to assume that this API will only be called in the polyfill case. It's a reasonable check for any node. Unfortunately, shadow roots have DOCUMENT_FRAGMENT type and thus difficult to determine.

Reply:
Yea I mean we could remove this function and just check tagName anywhere we care. Or rename to `isPolyfillShadowRoot`. Save a few bytes for an unused code path. :)